File: CarHomeWidget/src/main/java/com/anod/car/home/appwidget/WidgetHelper.java
package com.anod.car.home.appwidget;

import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Context;

/**
 * @author alex
 * @date 5/24/13
 */
public class WidgetHelper {

    static final String PROVIDER_DEFAULT = "com.anod.car.home.LargeProvider";
    private static final String PROVIDER_SHORTCUT = "com.anod.car.home.appwidget.ShortcutProvider";

    /**
     * Build {@link ComponentName} describing this specific
     * {@link android.appwidget.AppWidgetProvider}
     */
    public static int[] getLargeWidgetIds(Context context) {
        AppWidgetManager wm = AppWidgetManager.getInstance(context);
        ComponentName provider = new ComponentName(context.getPackageName(), PROVIDER_DEFAULT);

        int[] appWidgetIds = wm.getAppWidgetIds(provider);
        return appWidgetIds;
    }

    public static int[] getShortcutWidgetIds(Context context) {
        AppWidgetManager wm = AppWidgetManager.getInstance(context);
        ComponentName provider = new ComponentName(context.getPackageName(), PROVIDER_SHORTCUT);

        int[] appWidgetIds = wm.getAppWidgetIds(provider);
        return appWidgetIds;
    }

    public static int[] getAllWidgetIds(Context context) {
        int[] array1 = getLargeWidgetIds(context);
        int[] array2 = getShortcutWidgetIds(context);

        int[] total = new int[array1.length + array2.length];
        System.arraycopy(array1, 0, total, 0, array1.length);
        System.arraycopy(array2, 0, total, array1.length, array2.length);

        return total;
    }

}

File: CarHomeWidget/src/main/java/com/anod/car/home/appwidget/WidgetHelper.kt
package com.anod.car.home.appwidget

import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context

import com.anod.car.home.LargeProvider

/**
 * @author alex
 * @date 5/24/13
 */
object WidgetHelper {

    /**
     * Build [ComponentName] describing this specific
     * [android.appwidget.AppWidgetProvider]
     */
    fun getLargeWidgetIds(context: Context): IntArray {
        val applicationContext = context.applicationContext
        val provider = ComponentName(applicationContext, LargeProvider::class.java)
        return AppWidgetManager.getInstance(applicationContext).getAppWidgetIds(provider)
    }

    fun getShortcutWidgetIds(context: Context): IntArray {
        val applicationContext = context.applicationContext
        val wm = AppWidgetManager.getInstance(applicationContext)
        val provider = ComponentName(applicationContext, ShortcutProvider::class.java)

        return wm.getAppWidgetIds(provider)
    }

    fun getAllWidgetIds(context: Context): IntArray {
        val array1 = getLargeWidgetIds(context)
        val array2 = getShortcutWidgetIds(context)

        val total = IntArray(array1.size + array2.size)
        System.arraycopy(array1, 0, total, 0, array1.size)
        System.arraycopy(array2, 0, total, array1.size, array2.size)

        return total
    }

}
